### PR TITLE
[GLIMMER] Dashed property names

### PR DESCRIPTION
**Pull request by chadhietala**
Sort of sad that this is even a feature. @krisselden @wycats or @chancancode do you have any suggestions on this, I didn't want to go weaving `isBlock` all over the place.

Part of #13644

### CODE REVIEW

File: packages/ember/tests/component_registration_test.js [hunk 1]
```diff
@@ -139,18 +139,6 @@ QUnit.test('Late-registered components can be rendered with ONLY the template re
   ok(!helpers['borf-snorlax'], 'Component wasn\'t saved to global helpers hash');
 });
 
-test('Component-like invocations are treated as bound paths if neither template nor component are registered on the container', function() {
```
Comment (chadhietala) [inline packages/ember/tests/component_registration_test.js]: Duplicated test.

**Comment by rwjblue** [comment]
This seems good to me...

**Comment by chancancode** [comment]
The syntax refinement in ember is getting pretty confusing, what about something like this?

```
diff --git a/packages/ember-glimmer/lib/environment.js b/packages/ember-glimmer/lib/environment.js
index ebe1df4..2c769f9 100644
--- a/packages/ember-glimmer/lib/environment.js
+++ b/packages/ember-glimmer/lib/environment.js
@@ -122,6 +122,13 @@ export default class Environment extends GlimmerEnvironment {
   }
 
   refineStatement(statement) {
+    // 1. resolve any native syntax – if, unless, with, each, and partial
+    let nativeSyntax = super.refineStatement(statement);
+
+    if (nativeSyntax) {
+      return nativeSyntax;
+    }
+
     let {
       isSimple,
       isInline,
@@ -133,37 +140,34 @@ export default class Environment extends GlimmerEnvironment {
       templates
     } = statement;
 
-    if (key !== 'partial' && isSimple && (isInline || isBlock)) {
+    if (isSimple && (isInline || isBlock)) {
+      // 2. built-in syntaxs
       if (key === 'component') {
         return new DynamicComponentSyntax({ args, templates });
       } else if (key === 'outlet') {
         return new OutletSyntax({ args });
+      }
+
+      // 3. resolve components
+      let internalKey = builtInComponents[key];
+      let definition = null;
+
+      if (internalKey) {
+        definition = this.getComponentDefinition([internalKey]);
       } else if (key.indexOf('-') >= 0) {
-        let definition = this.getComponentDefinition(path);
-
-        if (definition) {
-          wrapClassBindingAttribute(args);
-          wrapClassAttribute(args);
-          return new CurlyComponentSyntax({ args, definition, templates });
-        } else if (isBlock && !this.hasHelper(key)) {
-          assert(`A helper named '${path[0]}' could not be found`, false);
-        }
-      } else {
-        // Check if it's a keyword
-        let mappedKey = builtInComponents[key];
-        if (mappedKey) {
-          let definition = this.getComponentDefinition([mappedKey]);
-          wrapClassBindingAttribute(args);
-          wrapClassAttribute(args);
-          return new CurlyComponentSyntax({ args, definition, templates });
-        }
+        definition = this.getComponentDefinition(path);
+      }
+
+      if (definition) {
+        wrapClassBindingAttribute(args);
+        wrapClassAttribute(args);
+        return new CurlyComponentSyntax({ args, definition, templates });
       }
     }
 
-    let nativeSyntax = super.refineStatement(statement);
-    assert(`Helpers may not be used in the block form, for example {{#${key}}}{{/${key}}}. Please use a component, or alternatively use the helper in combination with a built-in Ember helper, for example {{#if (${key})}}{{/if}}.`, !nativeSyntax && key && this.hasHelper(key) ? !isBlock : true);
-    assert(`Helpers may not be used in the element form.`, !nativeSyntax && key && this.hasHelper(key) ? !isModifier : true);
-    return nativeSyntax;
+    assert(`Helpers may not be used in the block form, for example {{#${key}}}{{/${key}}}. Please use a component, or alternatively use the helper in combination with a built-in Ember helper, for example {{#if (${key})}}{{/if}}.`, !isBlock || !this.hasHelper(key));
+    assert(`Helpers may not be used in the element form.`, !isModifier || !this.hasHelper(key));
+    assert(`Could not find component named "${key}" (no component or template with that name was found)`, isBlock);
   }
 
   hasComponentDefinition() {
@@ -179,8 +183,6 @@ export default class Environment extends GlimmerEnvironment {
 
       if (ComponentClass || layout) {
         definition = this._components[name] = new CurlyComponentDefinition(name, ComponentClass, layout);
-      } else if (!this.hasHelper(name)) {
-        assert(`Glimmer error: Could not find component named "${name}" (no component or template with that name was found)`, !!(ComponentClass || layout));
       }
     }
 ```

**Comment by chadhietala** [comment]
@chancancode I tried this locally and all of the late bound layout stuff is going to fail.

File: packages/ember-glimmer/lib/environment.js [hunk 2]
```diff
@@ -133,37 +170,37 @@ export default class Environment extends GlimmerEnvironment {
       templates
     } = statement;
 
-    if (key !== 'partial' && isSimple && (isInline || isBlock)) {
+    assert(`You attempted to overwrite the built-in syntax "${key}" which is not allowed. Please rename the helper.`, !(builtInHelpers[key] && this.owner.hasRegistration(`helper:${key}`)));
```
Comment (chancancode) [inline packages/ember-glimmer/lib/environment.js]: s/syntax/helper

File: packages/ember-glimmer/lib/syntax/dynamic-component.js [hunk 3]
```diff
@@ -10,6 +11,8 @@ function dynamicComponentFor(vm) {
     let name = nameRef.value();
     let definition = env.getComponentDefinition([name]);
 
+    assert(`Glimmer error: Could not find component named "${name}" (no component or template with that name was found)`, definition);
```
Comment (chancancode) [inline packages/ember-glimmer/lib/syntax/dynamic-component.js]: I think we can drop the "Glimmer error" prefix here too

File: packages/ember-glimmer/lib/syntax/dynamic-component.js [hunk 4]
```diff
@@ -44,6 +47,9 @@ class DynamicComponentReference {
 
     if (typeof name === 'string') {
       let definition = env.getComponentDefinition([name]);
+
+      assert(`Glimmer error: Could not find component named "${name}" (no component or template with that name was found)`, definition);
```
Comment (chancancode) [inline packages/ember-glimmer/lib/syntax/dynamic-component.js]: ditto

**Comment by chancancode** [comment]
LGTM 👍

**Comment by krisselden** [comment]
There was a problem with installing node earlier, I refreshed Travis. The only failure now is linting. You added a block comment and didn't specify @private or @public

**Comment by chadhietala** [comment]
Should be good to go now.

**Pull request merged**